### PR TITLE
[2d] Ensure gradient clear out pending AsyncCollectiveTensor in FSDP Extension

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116573
* #116559
* #116426
* #116244
* __->__ #116122

As titled, this PR adds gradient hook to the FSDP DTensor extension, to check if there's gradients that are AsyncCollectiveTensors, if there're some, we eagerly wait there.

This is needed because sometimes the parameter's gradient might still pending with AsyncCollectiveTensor, if we directly feed them to FSDP then FSDP would use the ACT's storage to do reduce_scatter, which is wrong.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_tp_integration.py [hunk 1]
```diff
@@ -314,6 +351,78 @@ def test_fsdp_tp_integration(self, tensor_parallel_size, cpu_offload):
         tp_fsdp_out = tp_fsdp_model(inp)
         self.assertEqual(fsdp_out, tp_fsdp_out)
 
+    @skip_if_lt_x_gpu(4)
+    def test_fsdp_tp_gradient_layout(self):
```
Comment (wanchaol) [inline test/distributed/fsdp/test_fsdp_tp_integration.py:355]: @awgu I added some tests on this change so this PR should be ready for review too :)

Comment (awgu) [reply test/distributed/fsdp/test_fsdp_tp_integration.py:355]: Thanks! I will review soon.

File: test/distributed/fsdp/test_fsdp_tp_integration.py [hunk 2]
```diff
@@ -314,6 +351,78 @@ def test_fsdp_tp_integration(self, tensor_parallel_size, cpu_offload):
...
+        # Shard with TP and test gradient
+        tp_model = parallelize_module(
+            model,
+            mesh_1d,
+            {
+                "mlp.net1": ColwiseParallel(),
+                "mlp.net2": RowwiseParallel(output_layouts=Shard(0)),
+            },
+        )
+        distribute_rmsnorm(tp_model.mlp_norm, mesh_1d)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_tp_integration.py:381]: IIUC, this means the RMSNorm runs as "sequence parallel"?

**Review by awgu — APPROVED** [review]
LGTM!

File: test/distributed/fsdp/test_fsdp_tp_integration.py (hunk 2 above)
Comment (wanchaol) [reply test/distributed/fsdp/test_fsdp_tp_integration.py:381]: Yep, this is sequence parallel layernorm/rmsnorm, the backward would produce partial dtensor, and the parameter would be replicated dtensor

**Review by fduwjj — APPROVED** [review]
LGTM

**Comment by wanchaol** [comment]
changed the purpose of this PR a bit to make the gradient hook to eagerly wait for the ACT.

File: torch/distributed/tensor/parallel/_data_parallel_utils.py [hunk 3]
```diff
@@ -1,24 +1,34 @@
...
 import torch
-from torch.distributed._tensor import DTensor as DistributedTensor
+from torch.distributed._functional_collectives import AsyncCollectiveTensor
+from torch.distributed._tensor import DTensor
 from torch.distributed._tensor.placement_types import DTensorSpec
 
 
+def sync_grad_hook(grad):
+    if isinstance(grad, AsyncCollectiveTensor):
+        grad.wait()
```
Comment (awgu) [inline torch/distributed/tensor/parallel/_data_parallel_utils.py:11]: @wanchaol Why do we discard the return value of `grad.wait()`, which unwraps the ACT? I.e., why not `grad = grad.wait()`?

Comment (wanchaol) [reply torch/distributed/tensor/parallel/_data_parallel_utils.py:11]: Good catch! I think we should unwrap yes, it's a bug. Could u help submit a fix on this?

Comment (awgu) [reply torch/distributed/tensor/parallel/_data_parallel_utils.py:11]: Oh wait, this looks like it was fixed on main already.
